### PR TITLE
Convert bytes to hex string with leading zeros

### PR DESCRIPTION
**Pull request by barthanssens**
Signed-off-by: Bart Hanssens <bart.hanssens@fedict.be>

This PR addresses GitHub issue: #775  .

Briefly describe the changes proposed in this PR:

* fixed bug (missing leading zeros)
* convert bytes to hexstring without BigInteger
* added javadoc and override

### CODE REVIEW

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java [hunk 1]
```diff
@@ -23,20 +22,31 @@
...
-			return "0" + result;
+
+		// convert to hexadecimal representation, with leading zeros
+		char[] hex = new char[hash.length * 2];
+		for (int i = 0, j = 0; i < hex.length; ) {
+			hex[i++] = HEX.charAt((hash[j] & 0xF0) >>> 4);
+			hex[i++] = HEX.charAt(hash[j++] & 0x0F);
 		}
-		return result;
+		return new String(hex);
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java]: As you said, there's many possible impls. Slightly worried that this approach is tricky to maintain. What do you think of this as an alternative fix:

     return String.format("%0" + (hash.length << 1) + "x", bi); 

(where `bi` is the `BigInteger` object from before the change)

Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java]: Also: can we add a unit test for this bug?

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java [hunk 2]
```diff
@@ -23,20 +22,31 @@
  * @author jeen
  */
 public abstract class HashFunction implements Function {
-
+private final String HEX = "0123456789abcdef";
```
Comment (ansell) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java]: Indentation

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java [hunk 3]
```diff
@@ -23,20 +22,31 @@
...
-		BigInteger bi = new BigInteger(1, hash);
-		String result = bi.toString(16);
-		if (result.length() % 2 != 0) {
-			return "0" + result;
+
+		// convert to hexadecimal representation, with leading zeros
+		char[] hex = new char[hash.length * 2];
+		for (int i = 0, j = 0; i < hex.length; ) {
+			hex[i++] = HEX.charAt((hash[j] & 0xF0) >>> 4);
+			hex[i++] = HEX.charAt(hash[j++] & 0x0F);
```
Comment (ansell) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java]: This seems hard to maintain compared to reusing one of the many existing implementations of hex encoding that exist.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java (hunk 1 above)
Comment (barthanssens) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/hash/HashFunction.java]: Ok, make senses to use the String.format and to add a unit test

**Comment by barthanssens** [comment]
Ok, fixed

**Pull request merged**